subprojects/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/ScriptPluginPluginResolver.java

```diff
@@ -0,0 +1,272 @@
+        pluginsLoader = new ScriptPluginLoaderClassLoader(textResourceLoader, targetScope.getExportClassLoader());
+        targetScope.createChild("script-plugins-loaders").export(pluginsLoader);
+    }
+
+    public static String getDescription() {
+        return "Script Plugins";
+    }
+
+    @Override
+    public void resolve(ContextAwarePluginRequest pluginRequest, PluginResolutionResult result) {
+        if (pluginRequest.getScript() == null) {
+            return;
+        }
+        if (pluginRequest.getModule() != null) {
+            throw new InvalidUserDataException("explicit artifact coordinates are not supported for script plugins applied using the plugins block");
+        }
+        if (pluginRequest.getVersion() != null) {
+            throw new InvalidUserDataException("explicit version is not supported for script plugins applied using the plugins block");
+        }
+        if (!pluginRequest.isApply()) {
+            throw new InvalidUserDataException("apply false is not supported for script plugins applied using the plugins block");
+        }
+
+        ScriptPluginImplementation scriptPluginImplementation = new ScriptPluginImplementation(pluginRequest, pluginsLoader);
+        result.found(getDescription(), new SimplePluginResolution(scriptPluginImplementation));
+    }
+
+    private static class ScriptPluginImplementation implements PluginImplementation<Object> {
+
+        private final ContextAwarePluginRequest pluginRequest;
+        private final ScriptPluginLoaderClassLoader pluginsLoader;
+
+        private Class<?> pluginLoaderClass;
+
+        private ScriptPluginImplementation(ContextAwarePluginRequest pluginRequest, ScriptPluginLoaderClassLoader pluginsLoader) {
+            this.pluginRequest = pluginRequest;
+            this.pluginsLoader = pluginsLoader;
+        }
+
+        @Override
+        public Class<?> asClass() {
+            if (pluginLoaderClass == null) {
+                pluginLoaderClass = pluginsLoader.defineScriptPluginLoaderClass(pluginRequest, getDisplayName());
+            }
+            return pluginLoaderClass;
+        }
+
+        @Override
+        public boolean isImperative() {
+            return true;
+        }
+
+        @Override
+        public boolean isHasRules() {
+            return false;
+        }
+
+        @Override
+        public Type getType() {
+            return Type.IMPERATIVE_CLASS;
+        }
+
+        @Override
+        public String getDisplayName() {
+            return "script plugin '" + pluginRequest.getRelativeScriptUri() + "'";
+        }
+
+        @Nullable
+        @Override
+        public PluginId getPluginId() {
+            return pluginRequest.getId();
+        }
+
+        @Override
+        public boolean isAlsoKnownAs(PluginId id) {
+            return id.equals(getPluginId());
+        }
+    }
+
+    private static class ScriptPluginLoaderClassLoader extends URLClassLoader {
```

Comment:
why do we need this?

Reply:
Because we're generating the byte code for the classes representing the script plugins in memory. But your comment made us realize we should switch to generating it on disk and use `ClassLoaderScope#export` instead.